client/driver/executor/executor.go

```diff
@@ -875,22 +877,51 @@ func NewLogRotatorWrapper(rotator *logging.FileRotator) (*logRotatorWrapper, err
 // called by the constructor and not the user of the wrapper.
 func (l *logRotatorWrapper) start() {
 	go func() {
-		io.Copy(l.rotatorWriter, l.processOutReader)
-		l.processOutReader.Close() // in case io.Copy stopped due to write error
-		close(l.hasFinishedCopied)
+		defer close(l.hasFinishedCopied)
+		_, err := io.Copy(l.rotatorWriter, l.processOutReader)
+		if err != nil {
+			// Close reader to propagate io error across pipe.
+			// Note that this may block until the process exits on
+			// Windows due to
+			// https://github.com/PowerShell/PowerShell/issues/4254
+			// or similar issues. Since this is already running in
+			// a goroutine its safe to block until the process is
+			// force-killed.
+			l.processOutReader.Close()
+		}
 	}()
 	return
 }
 
 // Close closes the rotator and the process writer to ensure that the Wait
 // command exits.
-func (l *logRotatorWrapper) Close() error {
+func (l *logRotatorWrapper) Close() {
 	// Wait up to the close tolerance before we force close
 	select {
 	case <-l.hasFinishedCopied:
 	case <-time.After(processOutputCloseTolerance):
 	}
-	err := l.processOutReader.Close()
+
+	// Closing the read side of a pipe may block on Windows if the process
```

Comment:
We added this close for processOutreader in https://github.com/hashicorp/nomad/pull/4359/commits/4150296b468c4193697dd25b748bdfb97fb7c749 to fix this [failing test|https://github.com/hashicorp/nomad/blob/8f4cbb215795353196b2d54da4f72f3d1eb3324c/client/driver/executor/executor_test.go#L96]. Removing this entirely doesn't affect that test (I tested with -count 50 locally). Wondering if we can remove that `l.processOutReader.Close()` line 893 entirely. That would simplify all this even more.

Reply:
My understanding of line 893 (now 890) is that it is to propagate errors from the io.Copy destination (`rotator`) back to the source (`processOutReader`).

So if `io.Copy` fails to write, it needs to signal to `processOutReader` that i will never be read again by `Close()`ing it.

I think this will handle cases like running-out-of-disk where we can no longer write anything, so we signal that to the process by closing its output (and likely causing the process to crash).